File: go/comparison/bychar.go
// Copyright 2000-2024 JetBrains s.r.o. and contributors. Use of this source code is governed by the Apache 2.0 license.
// Ported from intellij-community to Go.

package comparison

import (
	"unicode/utf8"

	"github.com/byron1st/file-diff/go/util"
)

// codePointsOffsets stores non-space (or punctuation) code points along with
// their byte offsets in the original string.
type codePointsOffsets struct {
	codePoints []int
	offsets    []int // byte offset of each code point in the original text
}

// charOffset returns the byte offset of the code point at index.
func (c *codePointsOffsets) charOffset(index int) int {
	return c.offsets[index]
}

// charOffsetAfter returns the byte offset after the code point at index.
func (c *codePointsOffsets) charOffsetAfter(index int) int {
	return c.offsets[index] + utf8.RuneLen(rune(c.codePoints[index]))
}

// CompareChars compares two strings at the character (code point) level.
func CompareChars(text1, text2 string) (FairDiffIterable, error) {
	cp1 := getAllCodePoints(text1)
	cp2 := getAllCodePoints(text2)

	iterable, err := Diff(cp1.codePoints, cp2.codePoints)
	if err != nil {
		return nil, err
	}

	builder := NewChangeBuilder(len(text1), len(text2))
	offset1, offset2 := 0, 0

	for _, r := range iterateAll(iterable) {
		end1 := offset1 + countCharBytes(cp1, r.rng.Start1, r.rng.End1)
		end2 := offset2 + countCharBytes(cp2, r.rng.Start2, r.rng.End2)
		if r.equal {
			builder.MarkEqualRange(offset1, offset2, end1, end2)
		}
		offset1 = end1
		offset2 = end2
	}

	return Fair(builder.Finish()), nil
}

// CompareCharsTwoStep compares two strings at the character level using a
// two-step approach: first match non-space characters, then fill gaps.
func CompareCharsTwoStep(text1, text2 string) (FairDiffIterable, error) {
	cp1 := getNonSpaceCodePoints(text1)
	cp2 := getNonSpaceCodePoints(text2)

	nonSpaceChanges, err := Diff(cp1.codePoints, cp2.codePoints)
	if err != nil {
		return nil, err
	}

	return matchAdjustmentSpaces(cp1, cp2, text1, text2, nonSpaceChanges)
}

// CompareCharsTrimWhitespaces compares two strings at the character level,
// trimming whitespace from change boundaries.
func CompareCharsTrimWhitespaces(text1, text2 string) (DiffIterable, error) {
	iterable, err := CompareCharsTwoStep(text1, text2)
	if err != nil {
		return nil, err
	}
	return newTrimSpacesCorrector(iterable, text1, text2).build(), nil
}

// CompareCharsIgnoreWhitespaces compares two strings ignoring whitespace.
func CompareCharsIgnoreWhitespaces(text1, text2 string) (DiffIterable, error) {
	cp1 := getNonSpaceCodePoints(text1)
	cp2 := getNonSpaceCodePoints(text2)

	changes, err := Diff(cp1.codePoints, cp2.codePoints)
	if err != nil {
		return nil, err
	}

	return matchAdjustmentSpacesIW(cp1, cp2, text1, text2, changes), nil
}

// ComparePunctuation compares only punctuation characters between two texts.
// All other characters are left unmatched.
func ComparePunctuation(text1, text2 string) (FairDiffIterable, error) {
	chars1 := getPunctuationChars(text1)
	chars2 := getPunctuationChars(text2)

	nonSpaceChanges, err := Diff(chars1.codePoints, chars2.codePoints)
	if err != nil {
		return nil, err
	}

	return transferPunctuation(chars1, chars2, text1, text2, nonSpaceChanges), nil
}

// matchAdjustmentSpaces converts a diff on non-space characters into a diff
// on the original texts by running a fair diff on gaps between matched characters.
func matchAdjustmentSpaces(
	cp1, cp2 *codePointsOffsets,
	text1, text2 string,
	changes FairDiffIterable,
) (FairDiffIterable, error) {
	return newDefaultCharChangeCorrector(cp1, cp2, text1, text2, changes).build()
}

// matchAdjustmentSpacesIW converts a diff on non-whitespace characters into a diff
// on the original texts. Matched characters include matched non-space characters
// plus all adjacent whitespace.
func matchAdjustmentSpacesIW(
	cp1, cp2 *codePointsOffsets,
	text1, text2 string,
	changes FairDiffIterable,
) DiffIterable {
	var ranges []util.Range

	for _, ch := range changes.Changes() {
		var startOffset1, endOffset1 int
		if ch.Start1 == ch.End1 {
			v := expandForwardW(cp1, cp2, text1, text2, ch, true)
			startOffset1, endOffset1 = v, v
		} else {
			startOffset1 = cp1.charOffset(ch.Start1)
			endOffset1 = cp1.charOffsetAfter(ch.End1 - 1)
		}

		var startOffset2, endOffset2 int
		if ch.Start2 == ch.End2 {
			v := expandForwardW(cp1, cp2, text1, text2, ch, false)
			startOffset2, endOffset2 = v, v
		} else {
			startOffset2 = cp2.charOffset(ch.Start2)
			endOffset2 = cp2.charOffsetAfter(ch.End2 - 1)
		}

		ranges = append(ranges, util.NewRange(startOffset1, endOffset1, startOffset2, endOffset2))
	}

	return CreateFromRanges(ranges, len(text1), len(text2))
}

// expandForwardW adjusts insertion/deletion placement to prefer matching whitespace.
func expandForwardW(
	cp1, cp2 *codePointsOffsets,
	text1, text2 string,
	ch util.Range,
	left bool,
) int {
	var offset1, offset2 int
	if ch.Start1 == 0 {
		offset1 = 0
	} else {
		offset1 = cp1.charOffsetAfter(ch.Start1 - 1)
	}
	if ch.Start2 == 0 {
		offset2 = 0
	} else {
		offset2 = cp2.charOffsetAfter(ch.Start2 - 1)
	}

	start := offset1
	if !left {
		start = offset2
	}

	return start + expandWhitespacesForward(text1, text2, offset1, offset2, len(text1), len(text2))
}

func transferPunctuation(
	chars1, chars2 *codePointsOffsets,
	text1, text2 string,
	changes FairDiffIterable,
) FairDiffIterable {
	builder := NewChangeBuilder(len(text1), len(text2))

	for _, r := range changes.Unchanged() {
		count := r.End1 - r.Start1
		for i := range count {
			// punctuation code points are always 1 byte in ASCII
			o1 := chars1.offsets[r.Start1+i]
			o2 := chars2.offsets[r.Start2+i]
			builder.MarkEqual(o1, o2)
		}
	}

	return Fair(builder.Finish())
}

// defaultCharChangeCorrector fills gaps between matched non-space code points
// by running a full character comparison on each gap.
type defaultCharChangeCorrector struct {
	cp1, cp2     *codePointsOffsets
	text1, text2 string
	changes      FairDiffIterable
}

func newDefaultCharChangeCorrector(
	cp1, cp2 *codePointsOffsets,
	text1, text2 string,
	changes FairDiffIterable,
) *defaultCharChangeCorrector {
	return &defaultCharChangeCorrector{cp1: cp1, cp2: cp2, text1: text1, text2: text2, changes: changes}
}

func (c *defaultCharChangeCorrector) build() (FairDiffIterable, error) {
	builder := NewChangeBuilder(len(c.text1), len(c.text2))
	last1, last2 := 0, 0

	for _, ch := range c.changes.Unchanged() {
		count := ch.End1 - ch.Start1
		for i := range count {
			start1 := c.cp1.charOffset(ch.Start1 + i)
			start2 := c.cp2.charOffset(ch.Start2 + i)
			end1 := c.cp1.charOffsetAfter(ch.Start1 + i)
			end2 := c.cp2.charOffsetAfter(ch.Start2 + i)

			if err := c.matchGap(builder, last1, start1, last2, start2); err != nil {
				return nil, err
			}
			builder.MarkEqualRange(start1, start2, end1, end2)

			last1 = end1
			last2 = end2
		}
	}
	if err := c.matchGap(builder, last1, len(c.text1), last2, len(c.text2)); err != nil {
		return nil, err
	}

	return Fair(builder.Finish()), nil
}

func (c *defaultCharChangeCorrector) matchGap(builder *ChangeBuilder, start1, end1, start2, end2 int) error {
	inner1 := c.text1[start1:end1]
	inner2 := c.text2[start2:end2]
	if len(inner1) == 0 && len(inner2) == 0 {
		return nil
	}

	innerChanges, err := CompareChars(inner1, inner2)
	if err != nil {
		return err
	}

	for _, ch := range innerChanges.Unchanged() {
		builder.MarkEqualCount(start1+ch.Start1, start2+ch.Start2, ch.End1-ch.Start1)
	}
	return nil
}

// iterateAllEntry represents a range that is either equal or changed.
type iterateAllEntry struct {
	rng   util.Range
	equal bool
}

// iterateAll yields all ranges (both changed and unchanged) in order.
func iterateAll(iterable DiffIterable) []iterateAllEntry {
	var result []iterateAllEntry
	changes := iterable.Changes()
	unchanged := iterable.Unchanged()

	ci, ui := 0, 0
	for ci < len(changes) || ui < len(unchanged) {
		takeChange := ci < len(changes) && (ui >= len(unchanged) ||
			changes[ci].Start1 < unchanged[ui].Start1 ||
			(changes[ci].Start1 == unchanged[ui].Start1 && changes[ci].Start2 < unchanged[ui].Start2))
		if takeChange {
			result = append(result, iterateAllEntry{rng: changes[ci], equal: false})
			ci++
		} else {
			result = append(result, iterateAllEntry{rng: unchanged[ui], equal: true})
			ui++
		}
	}
	return result
}

// getAllCodePoints extracts all code points from text along with byte offsets.
func getAllCodePoints(text string) *codePointsOffsets {
	var cps []int
	var offsets []int
	for offset, r := range text {
		cps = append(cps, int(r))
		offsets = append(offsets, offset)
	}
	return &codePointsOffsets{codePoints: cps, offsets: offsets}
}

// getNonSpaceCodePoints extracts non-whitespace code points with byte offsets.
func getNonSpaceCodePoints(text string) *codePointsOffsets {
	var cps []int
	var offsets []int
	for offset, r := range text {
		if !isWhiteSpaceCodePoint(r) {
			cps = append(cps, int(r))
			offsets = append(offsets, offset)
		}
	}
	return &codePointsOffsets{codePoints: cps, offsets: offsets}
}

// getPunctuationChars extracts punctuation characters with byte offsets.
func getPunctuationChars(text string) *codePointsOffsets {
	var cps []int
	var offsets []int
	for i, r := range text {
		if isPunctuation(r) {
			cps = append(cps, int(r))
			offsets = append(offsets, i)
		}
	}
	return &codePointsOffsets{codePoints: cps, offsets: offsets}
}

// countCharBytes counts total byte length of code points in range [start, end).
func countCharBytes(cp *codePointsOffsets, start, end int) int {
	total := 0
	for i := start; i < end; i++ {
		total += utf8.RuneLen(rune(cp.codePoints[i]))
	}
	return total
}

File: swift/Sources/FileDiff/ByChar.swift
// Copyright 2000-2025 JetBrains s.r.o. and contributors. Use of this source code is governed by the Apache 2.0 license.
// Ported from intellij-community to Swift (via the Go port in this repository).

import Foundation

/// Compares two texts at the Unicode code-point level and returns a diff
/// expressed in UTF-8 byte offsets.
public func compareChars(_ text1: String, _ text2: String) -> FairDiffIterable {
    let bytes1 = Array(text1.utf8)
    let bytes2 = Array(text2.utf8)
    let cp1 = getAllCodePoints(text1)
    let cp2 = getAllCodePoints(text2)

    let iterable = diff(cp1.codePoints, cp2.codePoints)
    let builder = ChangeBuilder(length1: bytes1.count, length2: bytes2.count)

    var offset1 = 0
    var offset2 = 0
    for entry in iterateAll(iterable) {
        let end1 = offset1 + countCharBytes(cp1, entry.range.left.lowerBound, entry.range.left.upperBound)
        let end2 = offset2 + countCharBytes(cp2, entry.range.right.lowerBound, entry.range.right.upperBound)
        if entry.equal {
            builder.markEqualRange(offset1, offset2, end1, end2)
        }
        offset1 = end1
        offset2 = end2
    }

    return fair(builder.finish())
}

/// Two-step character comparison: matches non-space characters first, then fills gaps.
public func compareCharsTwoStep(_ text1: String, _ text2: String) -> FairDiffIterable {
    let cp1 = getNonSpaceCodePoints(text1)
    let cp2 = getNonSpaceCodePoints(text2)
    let nonSpaceChanges = diff(cp1.codePoints, cp2.codePoints)
    return matchAdjustmentSpaces(cp1: cp1, cp2: cp2, text1: text1, text2: text2, changes: nonSpaceChanges)
}

/// Character-level comparison that trims whitespace from change boundaries.
public func compareCharsTrimWhitespaces(_ text1: String, _ text2: String) -> DiffIterable {
    let iterable = compareCharsTwoStep(text1, text2)
    let bytes1 = Array(text1.utf8)
    let bytes2 = Array(text2.utf8)
    return matchAdjustmentWhitespaces(
        bytes1: bytes1, bytes2: bytes2,
        iterable: iterable, policy: .trimWhitespaces
    )
}

/// Character-level comparison that ignores whitespace differences.
public func compareCharsIgnoreWhitespaces(_ text1: String, _ text2: String) -> DiffIterable {
    let bytes1 = Array(text1.utf8)
    let bytes2 = Array(text2.utf8)
    let cp1 = getNonSpaceCodePoints(text1)
    let cp2 = getNonSpaceCodePoints(text2)
    let changes = diff(cp1.codePoints, cp2.codePoints)
    return matchAdjustmentSpacesIW(
        cp1: cp1, cp2: cp2,
        bytes1: bytes1, bytes2: bytes2,
        changes: changes
    )
}

func comparePunctuation(_ text1: String, _ text2: String) -> FairDiffIterable {
    let bytes1 = Array(text1.utf8)
    let bytes2 = Array(text2.utf8)
    let chars1 = getPunctuationChars(text1)
    let chars2 = getPunctuationChars(text2)
    let nonSpaceChanges = diff(chars1.codePoints, chars2.codePoints)
    return transferPunctuation(
        chars1: chars1, chars2: chars2,
        bytes1: bytes1, bytes2: bytes2,
        changes: nonSpaceChanges
    )
}

// MARK: - Code point extraction

struct CodePointsOffsets {
    let codePoints: [Int]
    let offsets: [Int]
}

func charOffset(_ cp: CodePointsOffsets, _ index: Int) -> Int {
    cp.offsets[index]
}

func charOffsetAfter(_ cp: CodePointsOffsets, _ index: Int) -> Int {
    cp.offsets[index] + utf8LengthForCodePoint(cp.codePoints[index])
}

func utf8LengthForCodePoint(_ value: Int) -> Int {
    if value < 0x80 {
        return 1
    }
    if value < 0x800 {
        return 2
    }
    if value < 0x10000 {
        return 3
    }
    return 4
}

func getAllCodePoints(_ text: String) -> CodePointsOffsets {
    var points: [Int] = []
    var offsets: [Int] = []
    var offset = 0
    for scalar in text.unicodeScalars {
        points.append(Int(scalar.value))
        offsets.append(offset)
        offset += utf8Length(scalar)
    }
    return CodePointsOffsets(codePoints: points, offsets: offsets)
}

func getNonSpaceCodePoints(_ text: String) -> CodePointsOffsets {
    var points: [Int] = []
    var offsets: [Int] = []
    var offset = 0
    for scalar in text.unicodeScalars {
        let size = utf8Length(scalar)
        if !CharacterUtils.isWhiteSpaceCodePoint(scalar) {
            points.append(Int(scalar.value))
            offsets.append(offset)
        }
        offset += size
    }
    return CodePointsOffsets(codePoints: points, offsets: offsets)
}

func getPunctuationChars(_ text: String) -> CodePointsOffsets {
    var points: [Int] = []
    var offsets: [Int] = []
    var offset = 0
    for scalar in text.unicodeScalars {
        let size = utf8Length(scalar)
        if CharacterUtils.isPunctuation(scalar) {
            points.append(Int(scalar.value))
            offsets.append(offset)
        }
        offset += size
    }
    return CodePointsOffsets(codePoints: points, offsets: offsets)
}

func countCharBytes(_ cp: CodePointsOffsets, _ start: Int, _ end: Int) -> Int {
    var total = 0
    for i in start..<end {
        total += utf8LengthForCodePoint(cp.codePoints[i])
    }
    return total
}

// MARK: - Iterate all ranges

struct IterateAllEntry {
    let range: DiffRange
    let equal: Bool
}

func iterateAll(_ iterable: DiffIterable) -> [IterateAllEntry] {
    let changes = iterable.changes
    let unchanged = iterable.unchanged
    var result: [IterateAllEntry] = []
    var ci = 0
    var ui = 0
    while ci < changes.count || ui < unchanged.count {
        let takeChange: Bool
        if ci >= changes.count {
            takeChange = false
        } else if ui >= unchanged.count {
            takeChange = true
        } else {
            let change = changes[ci]
            let uc = unchanged[ui]
            if change.left.lowerBound < uc.left.lowerBound {
                takeChange = true
            } else if change.left.lowerBound == uc.left.lowerBound, change.right.lowerBound < uc.right.lowerBound {
                takeChange = true
            } else {
                takeChange = false
            }
        }
        if takeChange {
            result.append(IterateAllEntry(range: changes[ci], equal: false))
            ci += 1
        } else {
            result.append(IterateAllEntry(range: unchanged[ui], equal: true))
            ui += 1
        }
    }
    return result
}

// MARK: - Adjustment

func matchAdjustmentSpaces(
    cp1: CodePointsOffsets, cp2: CodePointsOffsets,
    text1: String, text2: String,
    changes: FairDiffIterable
) -> FairDiffIterable {
    DefaultCharChangeCorrector(
        cp1: cp1, cp2: cp2,
        text1: text1, text2: text2,
        changes: changes
    ).build()
}

func matchAdjustmentSpacesIW(
    cp1: CodePointsOffsets, cp2: CodePointsOffsets,
    bytes1: [UInt8], bytes2: [UInt8],
    changes: FairDiffIterable
) -> DiffIterable {
    var ranges: [DiffRange] = []
    for change in changes.changes {
        let startOffset1: Int
        let endOffset1: Int
        if change.left.lowerBound == change.left.upperBound {
            let value = expandForwardW(cp1: cp1, cp2: cp2, bytes1: bytes1, bytes2: bytes2, change: change, left: true)
            startOffset1 = value
            endOffset1 = value
        } else {
            startOffset1 = charOffset(cp1, change.left.lowerBound)
            endOffset1 = charOffsetAfter(cp1, change.left.upperBound - 1)
        }

        let startOffset2: Int
        let endOffset2: Int
        if change.right.lowerBound == change.right.upperBound {
            let value = expandForwardW(cp1: cp1, cp2: cp2, bytes1: bytes1, bytes2: bytes2, change: change, left: false)
            startOffset2 = value
            endOffset2 = value
        } else {
            startOffset2 = charOffset(cp2, change.right.lowerBound)
            endOffset2 = charOffsetAfter(cp2, change.right.upperBound - 1)
        }

        ranges.append(DiffRange(
            start1: startOffset1, end1: endOffset1,
            start2: startOffset2, end2: endOffset2
        ))
    }
    return createFromRanges(ranges, length1: bytes1.count, length2: bytes2.count)
}

// swiftlint:disable function_parameter_count
private func expandForwardW(
    cp1: CodePointsOffsets, cp2: CodePointsOffsets,
    bytes1: [UInt8], bytes2: [UInt8],
    change: DiffRange,
    left: Bool
) -> Int {
    // swiftlint:enable function_parameter_count
    let offset1 = change.left.lowerBound == 0 ? 0 : charOffsetAfter(cp1, change.left.lowerBound - 1)
    let offset2 = change.right.lowerBound == 0 ? 0 : charOffsetAfter(cp2, change.right.lowerBound - 1)
    let start = left ? offset1 : offset2
    let shift = TrimUtils.expandWhitespacesForward(
        bytes1, bytes2,
        offset1, offset2,
        bytes1.count, bytes2.count
    )
    return start + shift
}

func transferPunctuation(
    chars1: CodePointsOffsets, chars2: CodePointsOffsets,
    bytes1: [UInt8], bytes2: [UInt8],
    changes: FairDiffIterable
) -> FairDiffIterable {
    let builder = ChangeBuilder(length1: bytes1.count, length2: bytes2.count)
    for unchanged in changes.unchanged {
        let count = unchanged.left.count
        for i in 0..<count {
            let o1 = chars1.offsets[unchanged.left.lowerBound + i]
            let o2 = chars2.offsets[unchanged.right.lowerBound + i]
            builder.markEqual(o1, o2)
        }
    }
    return fair(builder.finish())
}

private final class DefaultCharChangeCorrector {
    private let cp1: CodePointsOffsets
    private let cp2: CodePointsOffsets
    private let text1: String
    private let text2: String
    private let bytes1: [UInt8]
    private let bytes2: [UInt8]
    private let changes: FairDiffIterable

    init(
        cp1: CodePointsOffsets, cp2: CodePointsOffsets,
        text1: String, text2: String,
        changes: FairDiffIterable
    ) {
        self.cp1 = cp1
        self.cp2 = cp2
        self.text1 = text1
        self.text2 = text2
        bytes1 = Array(text1.utf8)
        bytes2 = Array(text2.utf8)
        self.changes = changes
    }

    func build() -> FairDiffIterable {
        let builder = ChangeBuilder(length1: bytes1.count, length2: bytes2.count)
        var last1 = 0
        var last2 = 0
        for change in changes.unchanged {
            let count = change.left.count
            for i in 0..<count {
                let start1 = charOffset(cp1, change.left.lowerBound + i)
                let start2 = charOffset(cp2, change.right.lowerBound + i)
                let end1 = charOffsetAfter(cp1, change.left.lowerBound + i)
                let end2 = charOffsetAfter(cp2, change.right.lowerBound + i)

                matchGap(builder: builder, start1: last1, end1: start1, start2: last2, end2: start2)
                builder.markEqualRange(start1, start2, end1, end2)
                last1 = end1
                last2 = end2
            }
        }
        matchGap(builder: builder, start1: last1, end1: bytes1.count, start2: last2, end2: bytes2.count)
        return fair(builder.finish())
    }

    private func matchGap(builder: ChangeBuilder, start1: Int, end1: Int, start2: Int, end2: Int) {
        if start1 == end1, start2 == end2 {
            return
        }
        let inner1 = String(bytes: Array(bytes1[start1..<end1]), encoding: .utf8) ?? ""
        let inner2 = String(bytes: Array(bytes2[start2..<end2]), encoding: .utf8) ?? ""
        let inner = compareChars(inner1, inner2)
        for change in inner.unchanged {
            builder.markEqualCount(
                start1 + change.left.lowerBound,
                start2 + change.right.lowerBound,
                change.left.count
            )
        }
    }
}
